rtos/ConditionVariable.h

```diff
@@ -0,0 +1,198 @@
+ * - Spurious notifications (not triggered by the application) can occur
+ *   and it is not defined when these occur.
+ *
+ * @note Synchronization level: Thread safe
+ *
+ * Example:
+ * @code
+ * #include "mbed.h"
+ *
+ * Mutex mutex;
+ * ConditionVariable cond(mutex);
+ *
+ * // These variables are protected by locking mutex
+ * uint32_t count = 0;
+ * bool done = false;
+ *
+ * void worker_thread()
+ * {
+ *     mutex.lock();
+ *     do {
+ *         printf("Worker: Count %lu\r\n", count);
+ *
+ *         // Wait for a condition to change
+ *         cond.wait();
+ *
+ *     } while (!done);
+ *     printf("Worker: Exiting\r\n");
+ *     mutex.unlock();
+ * }
+ *
+ * int main() {
+ *     Thread thread;
+ *     thread.start(worker_thread);
+ *
+ *     for (int i = 0; i < 5; i++) {
+ *
+ *         mutex.lock();
+ *         // Change count and notify waiters of this
+ *         count++;
+ *         printf("Main: Set count to %lu\r\n", count);
+ *         cond.notify_all();
+ *         mutex.unlock();
+ *
+ *         wait(1.0);
+ *     }
+ *
+ *     mutex.lock();
+ *     // Change done and notify waiters of this
+ *     done = true;
+ *     printf("Main: Set done\r\n");
+ *     cond.notify_all();
+ *     mutex.unlock();
+ *
+ *     thread.join();
+ * }
+ * @endcode
+ */
+class ConditionVariable : private mbed::NonCopyable<ConditionVariable> {
+public:
+    /** Create and Initialize a ConditionVariable object */
+    ConditionVariable(Mutex &mutex);
+
+    /** Wait for a notification
+     *
+     * Wait until a notification occurs.
+     *
+     * @note - The thread calling this function must be the owner of the
+     * ConditionVariable's mutex and it must be locked exactly once
+     * @note - Spurious notifications can occur so the caller of this API
+     * should check to make sure the condition they are waiting on has
+     * been met
+     *
+     * Example:
+     * @code
+     * mutex.lock();
+     * while (!condition_met) {
+     *     cond.wait();
+     * }
+     * mutex.unlock();
+     * @endcode
+     */
+    void wait();
+
+    /** Wait for a notification or timeout
+     *
+     * @param   millisec  timeout value or 0 in case of no time-out.
+     * @return  true if a timeout occurred, false otherwise.
+     *
+     * @note - The thread calling this function must be the owner of the
+     * ConditionVariable's mutex and it must be locked exactly once
+     * @note - Spurious notifications can occur so the caller of this API
+     * should check to make sure the condition they are waiting on has
+     * been met
+     *
+     * Example:
+     * @code
+     * mutex.lock();
+     * Timer timer;
+     * timer.start();
+     *
+     * bool timed_out = false;
+     * uitn32_t time_left = TIMEOUT;
+     * while (!condition_met && !timed_out) {
+     *     timed_out = cond.wait_for(time_left);
+     *     uint32_t elapsed = timer.read_ms();
+     *     time_left = elapsed > TIMEOUT ? 0 : TIMEOUT - elapsed;
+     * }
```

Comment:
Having an example helps, but in a way it becomes an illustration of the API deficiency.

This example enters the problem area came up in EventQueue recently - use of Timer being bad for power, and the oddity of using twin timing sources: https://github.com/ARMmbed/mbed-os/pull/5328 - the consensus there led to https://github.com/ARMmbed/mbed-os/issues/5378 = "EventQueue should use RTOS tick count"

Finding it hard to resist the urge to rewrite this example to avoid Timer and use osKernelGetTickCount.

But then, if we could use osKernelGetTickCount, we might as well just have `wait_until` using it.

I think I need to go back to your previous statement:

> Currently there aren't any C++ RTOS APIs to get the kernel tick. Because of this [`wait_until`] couldn't be used with the C++ RTOS API alone. This would be better off as a future addition when a way to get the tick count from the C++ RTOS API is available.

Do we really need every API to be in C++? Why can't we expect people to use some CMSIS-RTOS APIs directly when doing RTOS-y things? Are we trying to add a complete mbed RTOS abstraction layer on top of the CMSIS-RTOS abstraction layer on top of RTX? How many abstraction layers do we need?

(If CMSIS-RTOS isn't there to be used, why is it in the system at all? You could have implemented all the mbed C++ RTOS abstractions directly on RTX).

For tick count in particular, it's a non-OO interface, so hard to see how one can improve on `osKernelGetTickCount()`. Will we gain anything by inventing `Kernel::get_tick_count()` or `Thread::get_tick_count()` methods?

If we do really want C++, can we pin down what is needed, to get it done quick?  I think these 3 would suffice:

 * `static uint32_t Thread::get_tick_count();` == `osKernelGetTickCount`
 * `static uint32_t Thread::get_tick_freq();` == `osKernelGetTickFrequency`
 * `static osStatus Thread::wait_until(uint32_t ticks);` == `osDelayUntil` (companion to existing `Thread::wait` == `osDelay`)
 * maybe milliseconds<->ticks helper functions (not directly present in CMSIS-RTOS)

Those could be added now to give access to the absolute time base without any bearing or statement on the future of the existing millisecond RTOS delay interfaces. It wouldn't matter whether the future of the RTOS delays was to retrospectively redefine them as ticks, or to say that they'll convert if tick frequency isn't 1000.

CC @pan- and @geky and @0xc0170 for more thoughts on this.